app/ui/browser/actions/main-actions.js

```diff
@@ -77,18 +99,54 @@ export function unbookmark(url) {
   };
 }
 
-export function navigatePageTo(pageId, location) {
-  return { type: types.NAVIGATE_PAGE_TO, pageId, location };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageTo(pageId, location, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(typeof location === 'string', '`location` must be a string.');
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    assert(getPageIndexById(getState(), pageId) >= 0,
+      `Page ${pageId} not found in current state.`);
+
+    dispatch({ type: types.NAVIGATE_PAGE_TO, pageId, location });
+    getWebViewById(doc, pageId).setAttribute('src', fixURL(location));
+  }, types.NAVIGATE_PAGE_TO);
 }
 
-export function navigatePageBack(pageId) {
-  return { type: types.NAVIGATE_PAGE_BACK, pageId };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageBack(pageId, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    const pageIndex = getPageIndexById(getState(), pageId);
+    assert(pageIndex >= 0, `Page ${pageId} not found in current state.`);
+    assert(getState().browserWindow.pages.get(pageIndex).canGoBack, 'Page cannot go back.');
+
+    dispatch({ type: types.NAVIGATE_PAGE_BACK, pageId });
+    getWebViewById(doc, pageId).goBack();
+  }, types.NAVIGATE_PAGE_BACK);
 }
 
-export function navigatePageForward(pageId) {
-  return { type: types.NAVIGATE_PAGE_FORWARD, pageId };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageForward(pageId, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    const pageIndex = getPageIndexById(getState(), pageId);
+    assert(pageIndex >= 0, `Page ${pageId} not found in current state.`);
+    assert(getState().browserWindow.pages.get(pageIndex).canGoForward, 'Page cannot go forward.');
+
+    dispatch({ type: types.NAVIGATE_PAGE_FORWARD, pageId });
+    getWebViewById(doc, pageId).goForward();
+  }, types.NAVIGATE_PAGE_FORWARD);
 }
 
-export function navigatePageRefresh(pageId) {
-  return { type: types.NAVIGATE_PAGE_REFRESH, pageId };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageRefresh(pageId, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    const pageIndex = getPageIndexById(getState(), pageId);
+    assert(pageIndex >= 0, `Page ${pageId} not found in current state.`);
+    assert(getState().browserWindow.pages.get(pageIndex).canRefresh, 'Page cannot refresh.');
+
+    dispatch({ type: types.NAVIGATE_PAGE_REFRESH, pageId });
```

Comment:
Correct; but not sure if that's the right way to go about it -- in that case, maybe it makes sense for the reducer to fire the side effect (webview navigation) for replayability?